dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,738 @@
+
+                    a, b, a_plus_b, a_minus_b = dali_server.produce_data(outs)
+
+        Example 3 - Full integration with PyTorch but using the original PyTorch DataLoader
+
+            .. code-block:: python
+
+                pipe = rn50_train_pipe(...)
+                with dali_proxy.DALIServer(pipe) as dali_server:
+                    dataset = torchvision.datasets.ImageFolder(
+                        jpeg, transform=dali_server.proxy, loader=read_filepath)
+
+                    # Using PyTorch DataLoader directly
+                    loader = torch.utils.data.DataLoader(
+                        dataset,
+                        batch_size=batch_size,
+                        num_workers=nworkers,
+                        drop_last=True,
+                    )
+
+                    for data, target in loader:
+                        # replaces the output reference with actual data
+                        data = dali_server.produce_data(data)
+                        ...
+        """
+        if not isinstance(pipeline, _Pipeline):
+            raise TypeError(f"Expected an NVIDIA DALI pipeline, got: {pipeline}")
+        else:
+            self._pipe = pipeline
+
+        # get the dali pipeline input names
+        self._dali_input_names = _external_source_node_names(self._pipe)
+        if len(self._dali_input_names) == 0:
+            raise RuntimeError("The provided pipeline doesn't have any inputs")
+
+        parameters = [Parameter("self", Parameter.POSITIONAL_OR_KEYWORD)]
+        for input_name in self._dali_input_names:
+            parameters.append(Parameter(input_name, Parameter.POSITIONAL_OR_KEYWORD))
+        return_annotation = tuple(DALIOutputSampleRef for _ in range(self._pipe.num_outputs))
+        self._signature = Signature(parameters, return_annotation=return_annotation)
+
+        # Multi-process queue used to transfer data from the pytorch workers to the main process
+        self._dali_input_q = _mp.Queue()
+        # Multi-process queue used by the main process to consume outputs from the DALI pipeline
+        self._dali_output_q = queue.Queue()
+        # Thread
+        self._thread = None
+        self._thread_stop_event = None
+        # Cache
+        self._cache_outputs = dict()
+        # Whether we want the order of DALI execution to be reproducible
+        self._deterministic = deterministic
+        # Proxy
+        self._proxy = None
+
+    def __del__(self):
+        self.stop_thread()
+
+    @property
+    def proxy(self):
+        if not self._proxy:
+
+            class _DALIProxyCallable(_DALIProxy):
+                def __init__(self, signature, dali_input_q, deterministic):
+                    super().__init__(signature, dali_input_q, deterministic)
+
+                @_modify_signature(self._signature)
+                def __call__(self, *args, **kwargs):
+                    return self._add_sample(self._signature.bind(self, *args, **kwargs))
```

Comment:
How about
```suggestion
                    bound_args = self._signature.bind(self, *args, **kwargs).arguments
                    return self._add_sample(bound_args)
```
...and now _add_sample just takes a dict.

Reply:
Done